### ISSUE TITLE
Fixer for line breaks around operators

### ISSUE THREAD

**Issue by EvgenyOrekhov** [PHP-CS-Fixer/PHP-CS-Fixer#3511]
Similar to <https://eslint.org/docs/rules/operator-linebreak>.

**Comment by Slamdunk** [comment]
Totally 👍 for this.

Developers who put operator at the end of line should be punished: either they love short meaningless variable names or scroll horizontally to understand the code :rage4:

**Comment by Slamdunk** [comment]
@SpacePossum whatsapp? You agree, don't you?

**Comment by kubawerlos** [comment]
> Developers who put operator at the end of line should be punished

With a year ban for programming, at least.

**Comment by SpacePossum** [comment]
I don't agree on developers being punished for not following a CSS-style. I wouldn't down vote a fixer for this, however there are a lot of FR's opened lately without the intent to develop those it seems.

**Comment by peter-gribanov** [comment]
> Developers who put operator at the end of line should be punished: either they love short meaningless variable names or scroll horizontally to understand the code :rage4:

In my opinion, exactly the opposite. Placing statements at the beginning of a line shifts the expression code relative to the first line and complicates reading the entire condition code. Using operators at the end of a line is a popular, well-established approach used in many languages.

See example [link](https://eslint.org/docs/rules/operator-linebreak) on ESLint from topic starter. The operator at the beginning is an exception, not a rule.

> Some developers find that placing operators at the beginning of the line makes the code more readable.

This is normal.


![image](https://user-images.githubusercontent.com/1954436/104935937-ea074200-59bc-11eb-89f4-c8dbe6d3a396.png)

This is controversial, but acceptable.


![image](https://user-images.githubusercontent.com/1954436/104935852-d0fe9100-59bc-11eb-8384-85addfca20f4.png)

This style makes reading difficult because of the offset of the condition code.


![image](https://user-images.githubusercontent.com/1954436/104936137-2aff5680-59bd-11eb-98ed-d0f7f50b826c.png)

**Comment by kubawerlos** [comment]
> This is normal.

> This is controversial, but acceptable.

> This style makes reading difficult because of the offset of the condition code.

First of all, those are your personal opinions. Someone can just switch the positions in your examples keeping the description.

As in the proverb: one rabbi will say “beginning”; another rabbi will say “end”.

There is no good or wrong style, just a matter of preference.

----

For me, the`beginning` version makes having code like this:
```
var foo = arg1
        + arg2
        - arg3
        + arg4;
```
more readable than:

```
var foo = arg1 +
          arg2 -
          arg3 +
          arg4;
```

**Comment by Slamdunk** [comment]
> In my opinion, exactly the opposite. Placing statements at the beginning of a line shifts the expression code relative to the first line and complicates reading the entire condition code. Using operators at the end of a line is a popular, well-established approach used in many languages.

Sorry for the rude tone I've used in the past, it wasn't my real intention.

Everything looks the same with short variable, but what about:

```php
$totalStockAmount = $this->gatherPreviousTotalStockAmountSnapshot() -
                    $this->gatherPurchasedProductCount() -
                    $this->calculateExpiredShortLivingProductsCount() +
                    $this->gatherReturnedInStockProductCountForRevision() -
                    $this->gatherBrokenProductsCount();

// vs

$totalStockAmount = $this->gatherPreviousTotalStockAmountSnapshot()
                  - $this->gatherPurchasedProductCount()
                  - $this->calculateExpiredShortLivingProductsCount()
                  + $this->gatherReturnedInStockProductCountForRevision()
                  - $this->gatherBrokenProductsCount();

```
In the first example I can barely keep track of what's happening, in the second one... a little better.

My company runs bank related calculation, and this kind of sum may have hundreds, yes, hundreds of operands, some in a 5 chars variable, other in a 200 chars gatherer method. Good luck finding a bug when the plus or minus are hidden by the horizontal scroll.

**Comment by peter-gribanov** [comment]
> Good luck finding a bug when the plus and minus are hidden by the horizontal scroll.

This is a sign of a violation of the coding style. No code should be hidden behind horizontal scrolling. All code must be available on the screen.

I admit that the use of arithmetic operators at the beginning of a line is justified if operators of different types are used.

In the case of logical operations, you always have the same operator and you just need to know which one and you don't need to look at the end of the line every time. If you need to use another logical operator in the same condition, then the conditions must be wrapped in parentheses. We focus on conditions, not operators.

A snippet of code from this project.

```php
if (
    // next Token is not a comment on its own line
   !($nextNonWhitespaceNestToken->isComment() && (
        !$tokens[$nextNonWhitespaceNestIndex - 1]->isWhitespace() ||
        !Preg::match('/\R/', $tokens[$nextNonWhitespaceNestIndex - 1]->getContent())
    )) &&
    // and it is not a `$foo = function () {};` situation
    !($nestToken->equals('}') && $nextNonWhitespaceNestToken->equalsAny([';', ',', ']', [CT::T_ARRAY_SQUARE_BRACE_CLOSE]])) &&
    // and it is not a `Foo::{bar}()` situation
    !($nestToken->equals('}') && $nextNonWhitespaceNestToken->equals('(')) &&
    // and it is not a `${"a"}->...` and `${"b{$foo}"}->...` situation
    !($nestToken->equals('}') && $tokens[$nestIndex - 1]->equalsAny(['"', "'", [T_CONSTANT_ENCAPSED_STRING], [T_VARIABLE]])) &&
    // and next token is not a closing tag that would break heredoc/nowdoc syntax
    !($tokens[$nestIndex - 1]->isGivenKind(T_END_HEREDOC) && $nextNonWhitespaceNestToken->isGivenKind(T_CLOSE_TAG))
) {
```

Now there is a problem with indentation which I mentioned earlier. This is not at all the same as what you are talking about and it does not make reading easier.

```php
$totalStockAmount = $this->gatherPreviousTotalStockAmountSnapshot()
    - $this->gatherPurchasedProductCount()
    - $this->calculateExpiredShortLivingProductsCount()
    + $this->gatherReturnedInStockProductCountForRevision()
    - $this->gatherBrokenProductsCount();

$result = $a
    && $b
    && $c;

if ($a
    && $b
    && $c
) {
}

$result = $a
    && (
        $b
        || $c
    )
    && $d;

// vs

$totalStockAmount = $this->gatherPreviousTotalStockAmountSnapshot()
                  - $this->gatherPurchasedProductCount()
                  - $this->calculateExpiredShortLivingProductsCount()
                  + $this->gatherReturnedInStockProductCountForRevision()
                  - $this->gatherBrokenProductsCount();

$result = $a
       && $b
       && $c;

if ($a
 && $b
 && $c
) {
}

$result = $a
       && (
           $b
        || $c
       )
       && $d;

// vs

$totalStockAmount = $this->gatherPreviousTotalStockAmountSnapshot();
$totalStockAmount -= $this->gatherPurchasedProductCount();
$totalStockAmount -= $this->calculateExpiredShortLivingProductsCount();
$totalStockAmount += $this->gatherReturnedInStockProductCountForRevision();
$totalStockAmount -= $this->gatherBrokenProductsCount();

$result =
    $a &&
    $b &&
    $c;

if ($a &&
    $b &&
    $c
) {
}

$result =
    $a &&
    (
        $b ||
        $c
    ) &&
    $d;
```

I do not condemn this approach. I'm trying to figure it out.

**Comment by Slamdunk** [comment]
```php
   !($nextNonWhitespaceNestToken->isComment() && (
        !$tokens[$nextNonWhitespaceNestIndex - 1]->isWhitespace() ||
        !Preg::match('/\R/', $tokens[$nextNonWhitespaceNestIndex - 1]->getContent())
    ))
```
This works only by convention, i.e. "open parenthesis only if you change the logical operator".
If you have a lot of developers to work with, conventions are a burden as they must be shared somehow.
My experience working with new devs tells me this generates less confusion and is quicker to adapt to:
```php
   !(
        $nextNonWhitespaceNestToken->isComment()
        && (
            !$tokens[$nextNonWhitespaceNestIndex - 1]->isWhitespace()
            || !Preg::match('/\R/', $tokens[$nextNonWhitespaceNestIndex - 1]->getContent())
        )
    )

**Comment by kubawerlos** [comment]
> No code should be hidden behind horizontal scrolling. All code must be available on the screen.

I envy you that the code you work with looks like this. In every project, I've worked the files are larger, some much larger. I cannot even imagine having this in unit tests - it almost never happens that I have test files that small.

BTW @SpacePossum is there a reason the feature PR was merged and this issue stayed opened?

**Comment by peter-gribanov** [comment]
> I envy you that the code you work with looks like this. In every project, I've worked the files are larger, some much larger. I cannot even imagine having this in unit tests - it almost never happens that I have test files that small.

Even before PSR, in the 90s, there were conventions about line length and method size. Of course, it is not always possible to follow this convention, but we must strive for it. If the line of code is too long, you might consider reformatting your code or introducing helper variables and methods. Methods that are too fat can signal that we need decomposition.

Unfortunately, this convention is poorly applied specifically in tests, since preparation of the environment and mocks takes up too much space, and it usually cannot be optimized.

**Comment by Wirone** [comment]
Fixed in #4021 🙂
